utils/sasldblistusers.8
.\" sasldblistusers - List users in sasldb file
.\" Tim Martin 3/8/00
.\"

.\" Copyright (c) 2000 Carnegie Mellon University.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\"
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer. 
.\"
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in
.\"    the documentation and/or other materials provided with the
.\"    distribution.
.\"
.\" 3. The name ""Carnegie Mellon University"" must not be used to
.\"    endorse or promote products derived from this software without
.\"    prior written permission. For permission or any other legal
.\"    details, please contact  
.\"      Office of Technology Transfer
.\"      Carnegie Mellon University
.\"      5000 Forbes Avenue
.\"      Pittsburgh, PA  15213-3890
.\"      (412) 268-4387, fax: (412) 268-7395
.\"      tech-transfer@andrew.cmu.edu
.\'
.\" 4. Redistributions of any form whatsoever must retain the following
.\"    acknowledgment:
.\"    ""This product includes software developed by Computing Services
.\"     at Carnegie Mellon University (http://www.cmu.edu/computing/).""
.\"
.\" CARNEGIE MELLON UNIVERSITY DISCLAIMS ALL WARRANTIES WITH REGARD TO
.\" THIS SOFTWARE, INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
.\" AND FITNESS, IN NO EVENT SHALL CARNEGIE MELLON UNIVERSITY BE LIABLE
.\" FOR ANY SPECIAL, INDIRECT OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\" WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN
.\" AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING
.\" OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.

.\"
.TH SASLDBLISTUSERS 8 "March 8, 2000" "CMU SASL"
.SH NAME
sasldblistusers \- list users in sasldb
.SH SYNOPSIS
.B sasldblistusers
.SH DESCRIPTION
.I sasldblistusers
is used to list the users in the SASL password database (usually
/etc/sasldb). This will NOT list all the users in /etc/passwd, shadow,
PAM, etc. only those created by SASL (usually by saslpasswd).
.SH OPTIONS
.TP
.SH KNOWN BUGS
This only works for sasldb files that use the gdbm library.
.TP
.SH SEE ALSO
rfc2222 \- Simple Authentication and Security Layer (SASL)


NEWS
New in 2.0.3-BETA
-----------------
* THIS IS A BETA-QUALITY RELEASE THAT IS ONLY INTENDED FOR USE BY
  DEVELOPERS WHOSE APPLICATIONS MAKE USE OF THE CYRUS SASL LIBRARY.
  IT *WILL BREAK* ANY APPLICATION EXPECTING THE SASLv1 API.
* This library should be fairly close to the core features that will be
  released in a final version of Cyrus SASLv2.  It probabally has bugs.
* Major new features included in this release:
  - The glue code now correctly handles client-send-first and server-send-last
    situations based on what the protocol and mechanism each support.
  - The sasldb code has been extracted from the main library and now resides
    in a separate libsasldb.la that is available at build time.
  - SASLdb now supports multiple auxilliary properties, though as distributed
    only userPassword is implemented and used.
  - Much improved configure checking for various items, including
    Berkeley DB, Kerberos, and GSSAPI.
  - Better (more standard) handling of realms in DIGEST-MD5.
  - A new Plugin Programmer's guide.
  - IPv6 support.
  - Error reporting now works in the GSSAPI plugin.
* See the ChangeLog for a more detailed list of changes.

New in 2.0.2-ALPHA
------------------
* THIS IS AN ALPHA-QUALITY RELEASE THAT IS ONLY INTENDED FOR DEVELOPERS
  WHOSE APPLICATIONS MAKE USE OF THE CYRUS SASL LIBRARY.  
* This release is intended to show developers that use Cyrus SASL what
  direction we are planning on taking the library so that they can make
  plans to migrate their applications accordingly
* Major new features included in this release:
  - Ability to compile a static library including all mechanisms.  This
    means lower memory usage and faster mechanism loading time, but
    is not for everyone (or even many people). See doc/advanced.html,
    as well as the '--with-staticsasl' configure flag.
  - Man pages should now all be present and are close to being correct.
  - Can now build libsfsasl and the smtptest program (using the --with-sfio
    configure flag)
  - Reverted to the v1 entry points for mechanisms, to allow v1 mechanisms
    to fail loading cleanly.
  - Auxprop and canon_user plugins can now load from DSOs
  - Java code now compiles (but is not well tested, or up to date with the
    current Java API draft)
  - Error handling and use of sasl_errdetail has been fleshed out and
    should now work in most cases.
* Still Coming:
  - Cleanup of the client-send-first and server-send-last situation
  - Error reporting in GSSAPI plugin
  - Move the sasldb code out of the main library and into plugins and
    utilities only.

New in 2.0.0-ALPHA
------------------
* THIS IS AN ALPHA-QUALITY RELEASE THAT IS ONLY INTENDED FOR DEVELOPERS
  WHOSE APPLICATIONS MAKE USE OF THE CYRUS SASL LIBRARY.  
* This release is intended to show developers that use Cyrus SASL what
  direction we are planning on taking the library so that they can make
  plans to migrate their applications accordingly
* This release implements the SASLv2 API.
  Some of the major improvements in the API include:
  - Memory management is now sane (whoever allocates the memory is responsible
    for freeing it)
  - Auxiliary Property plugin support (ability to interface with directory
    services as part of authentication)
  - Username canonification plugin support
  - Improved error reporting (not fully implemented in this release)
  - Database support has been simplified.  We now maintain only a single
    store of plaintext passwords that is shared by all supplied plugins
    (using the auxiliary property interface).
  The new API is more fully documented in the header files sasl.h, saslplug.h
  saslutil.h, and prop.h.  The man pages, programmers guide, and system
  administrators guide have also been rewritten to deal with the new API.
* There is still a good amount of work to be done, and as this code is alpha
  quality, it has bugs, known and unknown.  Please either use our bugzilla at
  http://bugzilla.andrew.cmu.edu, or email cyrus-bugs@andrew.cmu.edu with
  questions, comments, or bug reports.
  - Most notably, the Java bindings have not been converted to work with
    the new API, and thus will not compile successfully.
  - The current development branch with this source is in our
    cvs repository as the "sasl-v2-rjs3" branch of the "sasl" collection.
    (see http://asg.web.cmu.edu/cyrus/download/anoncvs.html for more info)


utils/sasldblistusers.c
/* sasldblistusers.c -- list users in sasldb
 * Tim Martin
 */
/* 
 * Copyright (c) 2001 Carnegie Mellon University.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer. 
 *
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in
 *    the documentation and/or other materials provided with the
 *    distribution.
 *
 * 3. The name "Carnegie Mellon University" must not be used to
 *    endorse or promote products derived from this software without
 *    prior written permission. For permission or any other legal
 *    details, please contact  
 *      Office of Technology Transfer
 *      Carnegie Mellon University
 *      5000 Forbes Avenue
 *      Pittsburgh, PA  15213-3890
 *      (412) 268-4387, fax: (412) 268-7395
 *      tech-transfer@andrew.cmu.edu
 *
 * 4. Redistributions of any form whatsoever must retain the following
 *    acknowledgment:
 *    "This product includes software developed by Computing Services
 *     at Carnegie Mellon University (http://www.cmu.edu/computing/)."
 *
 * CARNEGIE MELLON UNIVERSITY DISCLAIMS ALL WARRANTIES WITH REGARD TO
 * THIS SOFTWARE, INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
 * AND FITNESS, IN NO EVENT SHALL CARNEGIE MELLON UNIVERSITY BE LIABLE
 * FOR ANY SPECIAL, INDIRECT OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN
 * AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING
 * OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

#include <config.h>

#include <stdio.h>
#include <stdlib.h>

#include <sasl.h>

typedef void *listcb_t(const char *, const char *);

void listusers_cb(const char *authid, const char *realm)
{
    if ( !authid || !realm) {
	fprintf(stderr,"userlist callback has bad param");
	return;
    }

    /* the entries that just say the mechanism exists */
    if (strlen(authid)==0) return;

    printf("user: %s realm: %s\n",authid,realm);
}

/*
 * List all users in database
 */

#ifdef SASL_GDBM

#include <gdbm.h>
#include <fcntl.h>
#include <sys/stat.h>

int listusers(const char *path, listcb_t *cb)
{
    GDBM_FILE indb, outdb;
    datum dkey, nextkey, ekey;

    indb = gdbm_open(path, 0, GDBM_READER, S_IRUSR | S_IWUSR, NULL);

    if (!indb) {
	fprintf(stderr, "can't open %s\n", path);
	return 1;
    }

    dkey = gdbm_firstkey(indb);

    while (dkey.dptr != NULL) {
	char *authid = dkey.dptr;
	char *realm  = dkey.dptr+strlen(authid)+1;

	if (*authid) {
	    /* don't check return values */
	    cb(authid,realm);
	}

	nextkey=gdbm_nextkey(indb, dkey);
	dkey=nextkey;
    }

    gdbm_close(indb);
}

#endif /* SASL_GDBM */

#ifdef SASL_NDBM

#include <ndbm.h>
#include <fcntl.h>
#include <sys/stat.h>

int listusers(const char *path, listcb_t *cb)
{
    DBM *indb;
    datum dkey, nextkey;

    indb = dbm_open(path, O_RDONLY, S_IRUSR | S_IWUSR);

    if (!indb) {
	fprintf(stderr, "can't open %s\n", path);
	return 1;
    }

    dkey = dbm_firstkey(indb);

    while (dkey.dptr != NULL) {
	char *authid = dkey.dptr;
	char *realm  = dkey.dptr+strlen(authid)+1;
	char *tmp    = realm + strlen(realm)+1;

	if (*authid) {
	    /* don't check return values */
	    cb(authid,realm);
	}

	nextkey=dbm_nextkey(indb);
	dkey=nextkey;
    }

    dbm_close(indb);
}

#endif /* SASL_NDBM */

#ifdef SASL_BERKELEYDB

#include <db.h>
/*
 * Open the database
 *
 */
static int berkeleydb_open(const char *path,DB **mbdb)
{
    int ret;

#if DB_VERSION_MAJOR < 3
    ret = db_open(path, DB_HASH, DB_CREATE, 0664, NULL, NULL, mbdb);
#else /* DB_VERSION_MAJOR < 3 */
    ret = db_create(mbdb, NULL, 0);
    if (ret == 0 && *mbdb != NULL)
    {
	    ret = (*mbdb)->open(*mbdb, path, NULL, DB_HASH, DB_CREATE, 0664);
	    if (ret != 0)
	    {
		    (void) (*mbdb)->close(*mbdb, 0);
		    *mbdb = NULL;
	    }
    }
#endif /* DB_VERSION_MAJOR < 3 */

    if (ret != 0) {
	fprintf(stderr,"Error opening password file %s\n", path);
	return SASL_FAIL;
    }

    return SASL_OK;
}

/*
 * Close the database
 *
 */

static void berkeleydb_close(DB *mbdb)
{
    int ret;
    
    ret = mbdb->close(mbdb, 0);
    if (ret!=0) {
	fprintf(stderr,"error closing sasldb: %s",
		strerror(ret));
    }
}

int listusers(const char *path, listcb_t *cb)
{
    int result;
    DB *mbdb = NULL;
    DBC *cursor;
    DBT key, data;

    /* open the db */
    result=berkeleydb_open(path, &mbdb);
    if (result!=SASL_OK) goto cleanup;

    /* make cursor */
#if DB_VERSION_MAJOR < 3
#if DB_VERSION_MINOR < 6
    result = mbdb->cursor(mbdb, NULL,&cursor); 
#else
    result = mbdb->cursor(mbdb, NULL,&cursor, 0); 
#endif /* DB_VERSION_MINOR < 7 */
#else /* DB_VERSION_MAJOR < 3 */
    result = mbdb->cursor(mbdb, NULL,&cursor, 0); 
#endif /* DB_VERSION_MAJOR < 3 */

    if (result!=0) {
	fprintf(stderr,"Making cursor failure: %s\n",strerror(result));
      result = SASL_FAIL;
      goto cleanup;
    }

    memset(&key,0, sizeof(key));
    memset(&data,0,sizeof(data));

    /* loop thru */
    result = cursor->c_get(cursor, &key, &data,
			   DB_FIRST);

    while (result != DB_NOTFOUND)
    {
	char *authid;
	char *realm;
	char realm_buf[2048];
	int numnulls = 0;
	unsigned int lup;

	/* make sure there is exactly 1 null */
	for (lup=0;lup<key.size;lup++)
	    if (((char *)key.data)[lup]=='\0')
		numnulls++;

	if (numnulls != 1) {
	    fprintf(stderr,"warning: probable database corruption\n");
	    result = cursor->c_get(cursor, &key, &data, DB_NEXT);
	    continue;
	}

	authid = key.data;
	realm  = authid + strlen(authid)+1;
	memcpy(realm_buf, realm, key.size - (strlen(authid)+1));
	realm_buf[key.size - (strlen(authid)+1)] = '\0';

	if (*authid) {
	    /* don't check return values */
	    cb(authid,realm_buf);
	}

	memset(&key,0, sizeof(key));
	memset(&data,0,sizeof(data));

	result = cursor->c_get(cursor, &key, &data, DB_NEXT);
    }

    if (result != DB_NOTFOUND) {
	fprintf(stderr,"failure: %s\n",strerror(result));
	result = SASL_FAIL;
	goto cleanup;
    }

    result = cursor->c_close(cursor);
    if (result!=0) result = SASL_FAIL;

    result = SASL_OK;

 cleanup:

    if (mbdb != NULL) berkeleydb_close(mbdb);
    return result;
}

#endif /* BERKELEY */

#if !defined(SASL_GDBM) && !defined(SASL_NDBM) && !defined(SASL_BERKELEYDB)
int listusers(const char *path, listcb_t *cb)
{
    fprintf(stderr,"Unsupported DB format");
    exit(1);
}

#endif /* BERKELEY */


int main(int argc, char **argv)
{
    char *db= SASL_DB_PATH;
    
    if (argc > 1)
	db = argv[1];

    listusers(db, (listcb_t *) &listusers_cb);

    exit(0);
}
